File: src/bin/0542-01-matrix.py
# @leetup=info id=542 lang=python3 slug=01-matrix



from typing import Optional

# @leetup=code
from collections import deque
import math


class Solution:
    def updateMatrix(self, mat: list[list[int]]) -> list[list[int]]:
        R = len(mat)
        C = len(mat[0])

        res = [[math.inf] * C for _ in range(R)]
        q = deque()

        for r in range(R):
            for c in range(C):
                if mat[r][c] == 0:
                    q.append(((r, c), 0))

        visited = set()
        while q:
            ((r, c), dist) = q.popleft()
            if not (0 <= r < R and 0 <= c < C):
                continue
            if (r, c) in visited:
                continue
            visited.add((r, c))
            res[r][c] = min(res[r][c], dist)
            q.append(((r, c + 1), dist + 1))
            q.append(((r, c - 1), dist + 1))
            q.append(((r + 1, c), dist + 1))
            q.append(((r - 1, c), dist + 1))

        return res


# @leetup=code

File: src/bin/0542-01-matrix.rs
// @leetup=info id=542 lang=rust slug=01-matrix

// @leetup=code
use std::collections::*;

impl Solution {
    pub fn update_matrix(mat: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let r_len = mat.len();
        let c_len = mat[0].len();

        let mut res = vec![vec![-1; c_len]; r_len];

        let mut q = VecDeque::new();
        for r in 0..r_len {
            for c in 0..c_len {
                if mat[r][c] == 0 {
                    q.push_back(((r, c), 0));
                }
            }
        }

        while let Some(((r, c), dist)) = q.pop_front() {
            if r < r_len && c < c_len {
                if res[r][c] == -1 {
                    res[r][c] = dist;
                    q.push_back(((r, c + 1), dist + 1));
                    q.push_back(((r + 1, c), dist + 1));
                    q.push_back(((r, c.wrapping_sub(1)), dist + 1));
                    q.push_back(((r.wrapping_sub(1), c), dist + 1));
                }
            }
        }

        res
    }
}
// @leetup=code

struct Solution;

fn main() {
    let _ = Solution::update_matrix;
}
